[6/20] overrides 2 (#74)
Contenda-Team/Contenda-landing-page: src/posts/overrides2.md

---
layout: "../layouts/BlogPost.astro"
title: "Play with more than just the prompts!"
slug: prompt-engineering-for-all
description: "Overrides 2: Electric Boogaloo"
author: "Johnson Lin"
added: "June 20 2023"
tags: ["announcements"]
---

> Last week we [announced the ability to override some of our default parameters with the API and released our prompts for inspiration](https://contenda.co/blog/prompt-engineering-for-all/). Read the full post to also get a tutorial on getting set up with the API and some initial experimentation with segment prompts. Check it out to get all caught up with getting started on experimenting with our text transformation. 

Reviewer: ```suggestion
Last week we [announced the ability to override some of our default parameters through the API](https://contenda.co/blog/prompt-engineering-for-all/) and released our prompts for inspiration. 
```

> Today we're following up on it and exploring some of the other parameters with more detail. We're going to be using the same video ([Nicholas Renotte's](https://www.youtube.com/@NicholasRenotte) video: [Building a Neural Network with PyTorch in 15 Minutes](https://www.youtube.com/watch?v=mozBidd58VQ)).

Reviewer: Should say what that detail is. Tell me why I should care upfront.

## Code Similarity Threshold

> The parameter `code_similarity_score_threshold` is something we use to trim down on redundant code snippets. After we extract all the text from a video and pull together all the code snippets, we check them against each other since often we're just building on top of the same code. If there's a significant overlap between adjacent code snippets (defined by the cosine similarity score threshold here), we just take the longest one with the assumption being that the longest one is most complete. 

Reviewer: ```suggestion
The parameter `code_similarity_score_threshold` is how trim down on redundant code snippets. After we extract all the text from a video and pull together all the code snippets, we check them against each other. If there's a significant overlap between adjacent code snippets (defined by the cosine similarity score threshold here), we just take the longest one with the assumption being that the longest one is most complete. 
```

The default parameter here is `0.5`, something we've settled on after some testing. But, let's change it to something much higher and try to include every code snippet we can:

```json
{
  "status_update_email": "johnson@contenda.co",
  "source_id": "youtube mozBidd58VQ",
  "type": "tutorial",
  "overrides": {
    "code_similarity_score_threshold": 0.99
  }
}
```

> This means that in order for an adjacent code snippet to be determined as redundant and not useful, it needs to have a cosine similarity score of `0.99` which would mean it's practically identical. So now we're passing every code snippet we've extracted from the video to GPT when generating a segment. The expected result would be that GPT includes more code snippets, some of which are redundant. Here's a segment from the default parameters: 

Reviewer: ```suggestion
We are now looking cosine similarity score of `0.99` which would mean it's practically identical. So now we're passing every code snippet we've extracted from the video to GPT when generating a segment. The expected result would be that GPT includes more code snippets, some of which are redundant. Here's a segment from the default parameters: 
```

Reviewer: i messed up just realized it should be "We are now looking for"

    for epoch in range(10):
        # Loop through the dataset
        for batch in dataset:
            # Training code goes here
```

Inside this loop, we'll include the necessary code to perform the forward pass, calculate the loss, perform backpropagation, and update the model weights.
````

Now, here's the same segment with the higher threshold: 

````markdown
To set up the training flow, we will use a standard Python function. First, include the typical `if __name__ == "__main__"` statement to ensure the following code runs only when this file is executed directly:

```python
if __name__ == "__main__":
```

Next, create a loop to iterate through the epochs. In this example, we'll train the model for 10 epochs:

```python
if __name__ == "__main__":
    num_epochs = 10
    for epoch in range(num_epochs):
        # Training code goes here
```

Inside the loop, we'll iterate through the batches in our dataset:

```python
if __name__ == "__main__":
    num_epochs = 10
    for epoch in range(num_epochs):
        for batch in dataset:
            # Batch processing code goes here
```

Now we're ready to write the training code for processing each batch.
````

> As expected, GPT is more likely to use all the code snippets which means continuously building upon a code snippet in the segment, rather than just give us the big one at the end. We found that this result to be a bit unnecessary and only made editing more tedious. But, if you'd prefer a result closer to this, you should try bumping this threshold higher than `0.5`. Or maybe you'd like to go the other way and cut down on code snippets even more, in which case you should try bumping this threshold lower.

Reviewer: ```suggestion
As expected, GPT is more likely to use all the code snippets which means continuously explaining, rather than just give us the explanation at the end. We found this result to be unnecessary and made editing more tedious. But, if you'd prefer a step-by-step explanation, you should try bumping this threshold higher than `0.5`. Or maybe you'd like to go the other way and cut down on code snippets even more, in which case you should try lowering this threshold.
```

```

It really follows it to a tee. Now, do you want every heading to look like `Heading 2: Electric Boogaloo`? That's a separate question, but at least now we know we can.

## Max Input Segment

The parameter `max_input_segment` refers to how big our transcript segment chunks are allowed to be. We use different ways to find what we think are natural places to split up a transcript. We break up a transcript since most videos are long enough that it's impossible to fit the whole transcript and have enough tokens to ask for a blog from GPT. 

Our default `max_input_segment` is 1500 tokens. If you increase it any further, you might not see any changes. This is because the models we have in place are what determine where we would want our transcript to be segmented and this number merely determines what is too long of a segment, not what is too short. But, we will see some change if we decrease it, since we do have mechanisms in place to split up our segments even further. The API call would look like this: 

```json
{
  "status_update_email": "johnson@contenda.co",
  "source_id": "youtube mozBidd58VQ",
  "type": "tutorial",
  "overrides": {
    "max_input_segment": 200
  }
}
```

> Our original blog has 21 segments. But, now with a much smaller max size for segments, we get 30 segments. Even from the introduction, we can see the change. What is 2 segments in the original blog:

Reviewer: ```suggestion
Our original blog has 21 segments. But, now with a much smaller max size for segments, we get 30 segments. Even from the introduction, we can see the change. The original blog's 2 segments becomes 3 segments in the new blog.

**Original:**
```

For this task, I set some rules to make it even more challenging. The primary constraint was the time limit of 15 minutes. Additionally, I wasn't allowed to look at any documentation or pre-existing code, nor could I use CodePilot (though my subscription had expired anyway). If I were to break any of these rules, a one-minute time penalty would be deducted from my total time limit. The goal was to build a deep neural network using PyTorch as quickly as possible under these conditions.
---
To raise the stakes for this challenge, I decided that if I failed to meet the 15-minute time limit, I would give away a $50 Amazon gift card. In the last two episodes, I didn't quite make the time limit, so I was eager to put in extra effort this time.

Before diving into the task, let's briefly discuss what PyTorch is. PyTorch is an open-source deep learning framework mainly used with Python. Developed by the team at Facebook, it accelerates the process of building deep learning models and is widely used in cutting-edge models and deep learning research. Now that we have a basic understanding of PyTorch, let's get started with the challenge!
```

> Becomes 3 segments in our new blog:

Reviewer: ```suggestion
**New:**
```

---
For this challenge, there were also stakes involved. If I failed to meet the 15-minute time limit, I would be giving away a $50 Amazon gift card to one of you. Considering the results from the last two episodes where the time limit wasn't reached, this attempt definitely required extra effort.

Before diving into the coding session, let's briefly discuss what PyTorch is. PyTorch is an open-source deep learning framework, primarily used with Python. Developed by the team at Facebook, it significantly accelerates the process of building deep learning models. PyTorch is widely used in cutting-edge models and deep learning research. With this background information, we were ready to embark on the challenge.
```

> Content wise, it's about the same. But it does get a bit lengthier when you have more segments. Our original blog comes out to around 3200 words. Our new blog comes out to around 3600 words. We prefer to keep our blogs a bit more concise, but definitely experiment with lowering the `max_input_size` if you want more content from GPT that you can edit down. 

Reviewer: ```suggestion
Content wise, it's about the same. But it does get lengthier when you have more segments. Our original blog comes out to around 3200 words. Our new blog comes out to around 3600 words. We prefer to keep our blogs a bit more concise, but definitely experiment with lowering the `max_input_size` if you want more content from GPT that you can edit down. 
```

---
layout: "../layouts/BlogPost.astro"
title: "Play with more than just the prompts!"
slug: overrides-pt2
description: "Overrides 2: Electric Boogaloo"
author: "Johnson Lin"
added: "June 20 2023"
tags: ["announcements"]
---

Last week we [announced the ability to override some of our default parameters through the API](https://contenda.co/blog/prompt-engineering-for-all/) and released our prompts for inspiration. 

> Today we're following up on it and exploring some of the other parameters with more detail so that you can better experiment and create a blog that fits your needs. We're going to be using the same video ([Nicholas Renotte's](https://www.youtube.com/@NicholasRenotte) video: [Building a Neural Network with PyTorch in 15 Minutes](https://www.youtube.com/watch?v=mozBidd58VQ)).

Reviewer: Not quite what I meant by "tell me why I should care". Users don't seek experimentation-- they seek an outcome. Describe the outcome.

## Code Similarity Threshold

The parameter `code_similarity_score_threshold` is how we trim down on redundant code snippets. After we extract all the text from a video and pull together all the code snippets, we check them against each other since often we're just building on top of the same code. If there's a significant overlap between adjacent code snippets (defined by the cosine similarity score threshold here), we just take the longest one with the assumption being that the longest one is most complete. 

> The default parameter here is `0.5`, something we've settled on after some testing. But, let's change it to something much higher and try to include every code snippet we can:

Reviewer: ```suggestion
The default parameter here is `0.5`. But let's change it to something much higher and try to include every code snippet we can:
```
